source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc

```diff
@@ -208,17 +211,15 @@ void GradientController::recordLatencySample(MonotonicTime rq_send_time) {
   const std::chrono::microseconds rq_latency =
       std::chrono::duration_cast<std::chrono::microseconds>(time_source_.monotonicTime() -
                                                             rq_send_time);
-  uint32_t sample_count;
   {
     absl::MutexLock ml(&sample_mutation_mtx_);
     hist_insert(latency_sample_hist_.get(), rq_latency.count(), 1);
-    sample_count = hist_sample_count(latency_sample_hist_.get());
   }
 
-  if (inMinRTTSamplingWindow() && sample_count >= config_.minRTTAggregateRequestCount()) {
-    // This sample has pushed the request count over the request count requirement for the minRTT
-    // recalculation. It must now be finished.
+  synchronizer_.syncPoint("pre_minrtt_update");
+  if (sample_mutation_mtx_.TryLock()) {
```

Comment:
It is unclear to me why sample_mutation_mtx_ is released after the hist_insert or why we use TryLock() here instead of an unconditional lock.

Should you consider doing the updateMinRTT() call just after the hist_insert above in order to avoid potential for starvation of updateMinRTT() in cases where there are many hist_inserts?

Reply:
Sorry, I was reasoning about this as if that was a separate mutex. Given the current way it's implemented, it may as well have been after the gist insert. 

The Trylock should be on a separate mutex, just for the minrtt update, for this to make sense. That minrtt update shouldn't be blocking all the worker threads from recording their latencies.

Reply:
I suggest doing the updateMinRTT just after hist_insert for now.  Some followup PR could attempt to split the mutexes.

Reply:
I think we have some options here regarding mutex:

1. use tryLock() for updateMinRTT() with same mutex (current approach)
  performance: better than before
  minRTTT accuracy: low -- cause substantial miss on updating minRTT,  when there are many hist_insert.
2. move updateMinRTT() right after hist_insert
  performance: slightly worse than before. --- contention, as the mutex involves larger code block
  minRTT accuracy: high
3. use tryLock() for updateMinRTT() with a separate mutex
  performance: better than before
  minRTT accuracy: mid - might cause to [miss updating minRTT in some cases](https://github.com/envoyproxy/envoy/pull/15004#discussion_r581315005), but subsequent requests should be able to update MinRTT.
4. use [double-checked lock](https://github.com/envoyproxy/envoy/pull/15004/commits/e09d0d3a18eefc68ae853019c78d3d6dcaa33e21)
  performance: same as before
  minRTT accuracy: high

If taking smaller code change as a factor, the double-check lock change seems pretty straightforward: we only need to add the check "inMinRTTSamplingWindow() && hist_sample_count(latency_sample_hist_.get()) >= config_.minRTTAggregateRequestCount()" after https://github.com/envoyproxy/envoy/blob/98ac59243cf0c2960f267664da88c08f8b9641d1/source/extensions/filters/http/adaptive_concurrency/controller/gradient_controller.cc#L116

Reply:
Let's just go with what @antoniovicente suggested and call it a day with this PR. 

Either you or I (if I have some free time) can follow up with a benchmark PR and see if any of this even matters. I suggest we just get this thing fixed and the PR closed.

Reply:
Sounds good to me, updated to move updateMinRTT() right after hist_insert.